### PR TITLE
feat: add support for varargs

### PR DESCRIPTION
**Pull request by magnus-madsen**
Example:

```scala
import java.nio.file.Path

@Test
def testInvokeVarArgs01(): Path \ IO =
    Path.of("hello", (...{}: Vector[String]))

@Test
def testInvokeVarArgs02(): Path \ IO =
    Path.of("hello", ...{"foo.txt"})

@Test
def testInvokeVarArgs03(): Path \ IO =
    Path.of("hello", ...{"foo.txt", "bar.txt"})
```

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze You may not like it... I don't even like it... But I want a solution for varargs which does not require a region.

I think this is a useful start. And we can iterate.

**Comment by mlutze** [comment]
I'm confused. Why are you calling a varargs function with an array? Why not use it as a varargs function?

**Comment by magnus-madsen** [comment]
> I'm confused. Why are you calling a varargs function with an array? Why not use it as a varargs function?

The JVM has no concept of varargs. Moreover, commons-lang3 has no concept of varargs.

**Comment by mlutze** [comment]
> > I'm confused. Why are you calling a varargs function with an array? Why not use it as a varargs function?
> 
> The JVM has no concept of varargs. Moreover, commons-lang3 has no concept of varargs.

The JVM has no concept of generics, either. Ihe JVM isn't our interface; whatever they put in JAR files is our interface.


That said, as long as we only allow putting literals in, I guess it's safe.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 1]
```diff
@@ -322,6 +322,17 @@ object TypeReduction {
       val arrType = phantomArr.getClass
       arrType
 
+    // Vectors
+    case Type.Apply(Type.Cst(TypeConstructor.Vector, _), elmType, _) =>
+      // Find the Java class of the element type.
+      val t = getJavaType(elmType)
+
+      // Find the Java class of the array type.
+      // See: https://stackoverflow.com/questions/1679421/how-to-get-the-array-class-for-a-given-class-in-java
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:331]: This is a gammeldags solution. Class arrayType is more direct. Maybe we can go through all instances of this in another PR

File: main/test/flix/Test.Exp.Jvm.InvokeMethod.VarArgs.flix [hunk 2]
```diff
@@ -0,0 +1,17 @@
+mod Test.Exp.Jvm.InvokeMethod.VarArgs {
+
+    import java.nio.file.Path
+
+    @Test
+    def testInvokeVarArgs01(): Path \ IO =
+        Path.of("hello", (...{}: Vector[String]))
```
Comment (mlutze) [inline main/test/flix/Test.Exp.Jvm.InvokeMethod.VarArgs.flix:7]: IMO it's best if we require a literal vector here, to ensure no one else owns the vector and depends on its immutability.

But I guess that does not work for the empty case.

**Review by mlutze — APPROVED** [review]

**Comment by magnus-madsen** [comment]
This PR is allows programmers to use ordinary Java constructors/methods with varargs with requiring them to learn/understand regions. Moreover, it provides some new syntax so there is a path towards a better implementation.

File: main/test/flix/Test.Exp.Jvm.InvokeMethod.VarArgs.flix (hunk 2 above)
Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Jvm.InvokeMethod.VarArgs.flix:7]: Sadly the common case of empty varargs is a case we cannot easily support :(

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:331]: See if you can get it to work. I could not!

**Pull request merged**